### PR TITLE
Add another api to create join plan

### LINKED ISSUE #4306: Add a new api to create join in LogicalPlanBuilder
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

When I try to add test cases, I find it is not easy to create join logical plan whose keys are expressions, because we have to create many additional projections. 
I would like to add a new api to create such join in LogicalPlanBuilder.

**Describe the solution you'd like**
Add a new api to create join in `LogicalPlanBuilder`. The api will be like:
```rust
 pub fn join_with_expr_keys(
        &self,
        right: &LogicalPlan,
        join_type: JoinType,
        join_keys: (Vec<impl Into<Expr>>, Vec<impl Into<Expr>>),
        filter: Option<Expr>,
    ) -> Result<Self> 
```

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by ygf11**
# Which issue does this PR close?

Closes #4306.

# Rationale for this change

I find it is not easy to create join logical plan whose keys are expressions, because we have to create many additional projections. 

This new method can reduce some boilerplate code.

# What changes are included in this PR?

* Introduce a `join_with_expr_keys` to LogicalPlanBuilder.
* Using this method when create join logical plan.
* Some unit tests.

# Are these changes tested?

Yes.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/logical_plan/builder.rs [hunk 1]
```diff
@@ -458,6 +461,81 @@ impl LogicalPlanBuilder {
...
+    /// Apply a join with on constraint which the key is an expression.
+    ///
+    /// Filter expression expected to contain non-equality predicates that can not be pushed
+    /// down to any of join inputs.
+    /// In case of outer join, filter applied to only matched rows.
+    pub fn join_with_expr_keys(
+        &self,
+        right: &LogicalPlan,
+        join_type: JoinType,
+        join_keys: (Vec<impl Into<Expr>>, Vec<impl Into<Expr>>),
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/builder.rs:473]: Given all the logic I think we now have in the optimizer (e.g. eliminate cross join https://github.com/apache/arrow-datafusion/pull/4185) I recommend we *ONLY* support specifying `filter` on this function and leave it up to the optimizer pass to sort out which are eqijoins and which are not

Comment (xudong963) [reply datafusion/expr/src/logical_plan/builder.rs:473]: I think so. -- We should "believe" our optimizer, If it doesn't perform well, we can polish it!

Comment (alamb) [reply datafusion/expr/src/logical_plan/builder.rs:473]: We should definitely test it, but I think the other passes will handle pushing predicates into Joins correctly -- for example https://github.com/apache/arrow-datafusion/blob/209c26670da4ee52cf13e4302250069a686fdaa2/datafusion/optimizer/src/filter_push_down.rs#L406-L443

Comment (ygf11) [reply datafusion/expr/src/logical_plan/builder.rs:473]: Thank you, got it. I will try it.

**Comment by alamb** [comment]
Closing as this PR is over a year old. Please feel free to reopen it / rebase it if you plan to keep working on it

**Pull request closed without merging**